# Write more functions for GTFS classes

Sunday, March 24, 2019

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd
import numpy as np

import os
import gtfs_transformer as gt

from shapely.geometry import Point
from shapely.ops import nearest_points
from geopy.distance import great_circle
import matplotlib.pyplot as plt
from tqdm import tqdm as tqdm_base, tqdm_notebook as tqdm

tqdm_base.pandas()

In [131]:
import copy

## Create GTFS objects for July 2018 and August 2018 by reading from directory

### To Do:

I'd like to add the ability to initialize the gtfs object directly from the `.zip` file downloaded from King County Metro, rather than having to unzip the file first. I don't think you can do this directly using [`pandas.read_csv`](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) because the `.zip` file contains multiple `.csv`'s. However, I think it should be possible using the Python [`zipfile`](https://pymotw.com/2/zipfile/) module.

In [117]:
#Initializes GTFS object with tables in given directories, inferring date from directory name
gtfs_jul18 = gt.StaticGTFS('../data/source/gtfs_20180718')
gtfs_aug18 = gt.StaticGTFS('../data/source/gtfs_20180815')

In [10]:
gtfs_aug18.table_names

['agency',
 'block',
 'block_trip',
 'calendar',
 'calendar_dates',
 'fare_attributes',
 'fare_rules',
 'routes',
 'shapes',
 'stop_times',
 'stops',
 'trips']

## Get route id's using `route_ids_from_names` function

In [13]:
# This creates a mapping for all routes in the routes table
route_ids = gtfs_aug18.route_ids_from_names()
route_ids.head(10)

route_id
route_short_name          
1                   100001
10                  100002
101                 100003
105                 100004
106                 100005
107                 100006
11                  100009
111                 100011
113                 100012
114                 100013

In [8]:
route_ids.loc['D Line']

route_id    102581
Name: D Line, dtype: int64

In [9]:
route_ids.loc[['10', 'D Line']]

route_id
route_short_name          
10                  100002
D Line              102581

In [12]:
# Can get same table by passing specific lines to function
gtfs_aug18.route_ids_from_names(10, 'D Line')

route_id
route_short_name          
10                  100002
D Line              102581

In [137]:
route_ids.at['10', 'route_id']

100002

In [139]:
# # This doesn't work...
# route_ids.at[['10', 'D Line'], 'route_id']

In [140]:
gtfs_aug18.route_ids_from_names(10, 'D Line').values

array([[100002],
       [102581]])

In [142]:
route_ids.head().to_dict()

{'route_id': {'1': 100001,
  '10': 100002,
  '101': 100003,
  '105': 100004,
  '106': 100005}}

In [144]:
route_ids.to_dict()['route_id']['10']

100002

In [21]:
route_ids.shape

(221, 1)

In [113]:
gtfs_aug18.routes.head(3)

route_id agency_id route_short_name  route_long_name  \
0    100001       KCM                1              NaN   
1    100002       KCM               10              NaN   
2    100003       KCM              101              NaN   

                                 route_desc  route_type  \
0                Kinnear - Downtown Seattle           3   
1           Capitol Hill - Downtown Seattle           3   
2  Renton Transit Center - Downtown Seattle           3   

                                           route_url  route_color  \
0  http://metro.kingcounty.gov/schedules/001/n0.html          NaN   
1  http://metro.kingcounty.gov/schedules/010/n0.html          NaN   
2  http://metro.kingcounty.gov/schedules/101/n0.html          NaN   

   route_text_color  
0               NaN  
1               NaN  
2               NaN

## We want to display trip headsign and `direction_id` for a route to know which direction is which

However, there may be more than one headsign for a given direction...

In [11]:
gtfs_aug18.trips.head()

route_id  service_id   trip_id      trip_headsign trip_short_name  \
0    100227       19845  34734386  Downtown Seattle            LOCAL   
1    100227       19845  34734387  Downtown Seattle            LOCAL   
2    100227       19845  34734388  Downtown Seattle            LOCAL   
3    100227       19845  34734389  Downtown Seattle            LOCAL   
4    100227       19845  34734390  Downtown Seattle            LOCAL   

   direction_id  block_id  shape_id  peak_flag  fare_id  
0             1   5045281  20047002          0    101.0  
1             1   5045281  20047002          0    101.0  
2             1   5045281  20047002          0    101.0  
3             1   5045281  20047002          0    101.0  
4             1   5045281  20047002          0    101.0

In [14]:
gtfs_aug18.trips.shape

(59857, 10)

In [150]:
gtfs_aug18.trips['trip_short_name'].unique()

array(['LOCAL', 'EXPRESS'], dtype=object)

In [151]:
# Less than 10% of trips are EXPRESS
gtfs_aug18.trips[gtfs_aug18.trips['trip_short_name']=='EXPRESS'].shape

(5195, 10)

### Try out different ways of aggregating the `trips` data

In [20]:
gtfs_aug18.trips.groupby(by=['route_id', 'trip_headsign']).agg({'direction_id': 'mean'}).shape

(542, 1)

In [26]:
gtfs_aug18.trips.groupby(by=['route_id', 'direction_id']).agg({'trip_headsign': 'max'}).shape

(438, 1)

In [28]:
gtfs_aug18.trips.groupby(by=['route_id', 'trip_headsign']).agg({'direction_id': 'max', 'shape_id': 'max'}).head(10)

direction_id  shape_id
route_id trip_headsign                                                    
100001   Downtown Seattle                                      1  20001010
         Kinnear Seattle Center W                              0  11001012
100002   Atlantic Base Via Broadway                            1  20010004
         Capitol Hill Via 15th Ave E                           0  11010010
         Downtown Seattle                                      1  20010007
100003   Downtown Seattle                                      1  10101008
         S Renton P&R                                          0  21101007
100004   Renton Highlands                                      0  11105002
         Renton Renton Tech Coll                               1  20105003
100005   International District Chinatown Station              1  10106037

In [24]:
gtfs_aug18.routes.shape

(221, 9)

In [25]:
gtfs_aug18.trips.direction_id.unique()

array([1, 0])

In [33]:
gtfs_aug18.trips.groupby(
    by=['route_id', 'trip_headsign', 'direction_id', 'shape_id']
).agg({'trip_id': ['count', 'min', 'max']}).head(20)

trip_id  \
                                                                                    count   
route_id trip_headsign                                      direction_id shape_id           
100001   Downtown Seattle                                   1            20001009       8   
                                                                         20001010     274   
         Kinnear Seattle Center W                           0            11001011     270   
                                                                         11001012      12   
100002   Atlantic Base Via Broadway                         1            20010004      10   
         Capitol Hill Via 15th Ave E                        0            11010005       8   
                                                                         11010006     470   
                                                                         11010010       8   
         Downtown Seattle                                   1            20010002     470   
                                                                         20010007       2   
100003   Downtown Seattle                                   1            10101006     114   
                                                                         10101008     114   
         S Renton P&R                                       0            21101005     121   
                                                                         21101007     121   
100004   Renton Highlands                                   0            11105002     168   
         Renton Renton Tech Coll                            1            20105003     174   
100005   International District Chinatown Station           1            10106037       6   
         International District Chinatown Station Rainie... 1            10106034     321   
         Renton Rainier Beach Station                       0            21106108     334   
100006   Beacon Hill Station Rainier Beach Station          1            10107018     246   

                                                                                             \
                                                                                        min   
route_id trip_headsign                                      direction_id shape_id             
100001   Downtown Seattle                                   1            20001009  34736146   
                                                                         20001010  34735996   
         Kinnear Seattle Center W                           0            11001011  34735973   
                                                                         11001012  34736010   
100002   Atlantic Base Via Broadway                         1            20010004  34740661   
         Capitol Hill Via 15th Ave E                        0            11010005  34735719   
                                                                         11010006  38426627   
                                                                         11010010  38426760   
         Downtown Seattle                                   1            20010002  34735660   
                                                                         20010007  34740670   
100003   Downtown Seattle                                   1            10101006  39170852   
                                                                         10101008  39566869   
         S Renton P&R                                       0            21101005  39170811   
                                                                         21101007  39566879   
100004   Renton Highlands                                   0            11105002  39169481   
         Renton Renton Tech Coll                            1            20105003  39169479   
100005   International District Chinatown Station           1            10106037  34816202   
         International District Chinatown 

In [30]:
gtfs_aug18.trips.groupby(
    by=['route_id', 'trip_headsign', 'shape_id', 'direction_id']
).agg({'trip_id': 'count'}).shape

(914, 1)

In [92]:
gtfs_aug18.trips.merge(gtfs_aug18.routes[['route_id', 'route_short_name', 'route_desc']], on='route_id').groupby(
    by=['route_id', 'route_short_name', 'route_desc', 'trip_headsign', 'direction_id', 'shape_id']
).agg({'trip_id': ['count', 'min', 'max']}).head(20)

trip_id  \
                                                                                                                                                       count   
route_id route_short_name route_desc                                        trip_headsign                                      direction_id shape_id           
100001   1                Kinnear - Downtown Seattle                        Downtown Seattle                                   1            20001009       8   
                                                                                                                                            20001010     274   
                                                                            Kinnear Seattle Center W                           0            11001011     270   
                                                                                                                                            11001012      12   
100002   10               Capitol Hill - Downtown Seattle                   Atlantic Base Via Broadway                         1            20010004      10   
                                                                            Capitol Hill Via 15th Ave E                        0            11010005       8   
                                                                                                                                            11010006     470   
                                                                                                                                            11010010       8   
                                                                            Downtown Seattle                                   1            20010002     470   
                                                                                                                                            20010007       2   
100003   101              Renton Transit Center - Downtown Seattle          Downtown Seattle                                   1            10101006     114   
                                                                                                                                            10101008     114   
                                                                            S Renton P&R                                       0            21101005     121   
                                                                                                                                            21101007     121   
100004   105              Renton Highlands - Renton Transit Center          Renton Highlands                                   0            11105002     168   
                                                                            Renton Renton Tech Coll                            1            20105003     174   
100005   106              Renton Transit Center - Skyway - Downtown Seattle International District Chinatown Station           1            10106037       6   
                                                                            International District Chinatown Station Rainie... 1            10106034     321   
                                                                            Renton Rainier Beach Station                       0            21106108     334   
100006   107              Renton Transit Center - Rainier Beach             Beacon Hill Station Rainier Beach Station          1            10107018     246   

                                                                                                                                                                \
                                                                                                                                                           min   
route_id route_short_name route_desc                                        trip_headsign                                      direction_id shap

### Ok, I think the following dataframe looks good for the final version

In [111]:
gtfs_aug18.trips.merge(
    gtfs_aug18.routes[['route_id', 'route_short_name', 'route_desc']],
    on='route_id'
).groupby(
    by=['route_short_name', 'direction_id']
).agg(
    {'route_desc': 'max', 
     'trip_headsign': lambda x: x.unique(),
     'shape_id': lambda x: len(x.unique()),
     'trip_id': 'count'}
).rename(
    columns={'shape_id': 'shape_count', 'trip_id': 'trip_count'}
        ).head(20)#.loc[('10',0), 'trip_headsign']
# .reset_index().set_index('route_short_name').head(20)

route_desc  \
route_short_name direction_id                                                      
1                0                                    Kinnear - Downtown Seattle   
                 1                                    Kinnear - Downtown Seattle   
10               0                               Capitol Hill - Downtown Seattle   
                 1                               Capitol Hill - Downtown Seattle   
101              0                      Renton Transit Center - Downtown Seattle   
                 1                      Renton Transit Center - Downtown Seattle   
102              0                                   Fairwood - Downtown Seattle   
                 1                                   Fairwood - Downtown Seattle   
105              0                      Renton Highlands - Renton Transit Center   
                 1                      Renton Highlands - Renton Transit Center   
106              0             Renton Transit Center - Skyway - Downtown Seattle   
                 1             Renton Transit Center - Skyway - Downtown Seattle   
107              0                         Renton Transit Center - Rainier Beach   
                 1                         Renton Transit Center - Rainier Beach   
11               0                Madison Park - Capitol Hill - Downtown Seattle   
                 1                Madison Park - Capitol Hill - Downtown Seattle   
111              0                              Lake Kathleen - Downtown Seattle   
                 1                              Lake Kathleen - Downtown Seattle   
113              0                                  Shorewood - Downtown Seattle   
                 1                                  Shorewood - Downtown Seattle   

                                                                   trip_headsign  \
route_short_name direction_id                                                      
1                0                                      Kinnear Seattle Center W   
                 1                                             Downtown Seattle    
10               0                                   Capitol Hill Via 15th Ave E   
                 1               [Downtown Seattle , Atlantic Base Via Broadway]   
101              0                                                 S Renton P&R    
                 1                                             Downtown Seattle    
102              0                        [Fairwood S Renton P&R, S Renton P&R ]   
                 1             [Downtown Seattle S Renton P&R, Downtown Seatt...   
105              0                                             Renton Highlands    
                 1                                       Renton Renton Tech Coll   
106              0                                  Renton Rainier Beach Station   
                 1             [International District Chinatown Station Rain...   
107              0                   Renton Transit Center Rainier Beach Station   
                 1                     Beacon Hill Station Rainier Beach Station   
11               0                                 Madison Park Via E Madison St   
                 1             [Downtown Seattle Via E Madison St, Stadium St...   
111              0                                        Lake Kathleen Via I-90   
                 1                                             Downtown Seattle    
113              0                                        Shorewood White Center   
                 1             [Downtown Seattle White Center, White Center S...   

                               shape_count  trip_count  
route_short_name direction_id                           
1                0                       2         282  
                 1                       2         282  
10               0                       3         486  
                 1                       3         482  
101              0            

In [110]:
# Hmm, using lambda x: x.unique() in .agg() returns a string if there's one object,
# and a numpy array of strings if there are more... Seems like that could be either useful or annoying...
# (Result of running the above cell with .loc[('10',0), 'trip_headsign'] uncommented)
type(_)

str

## Write a function to return a summary dataframe like the one above, optionally passing in specific route names

Note that the `Redmond LOOP` line only has one direction. The length of the returned dataframe for all routes is 438, which is 4 less than twice the number of routes (221), presumably because some routes only have one direction, like the Redmond Loop.

In [210]:
gtfs_aug18.trips_by_route_and_direction(10, 7, 'C Line', 'E Line', 8, 'Redmond LOOP', 41)

route_desc  \
route_short_name direction_id                                                     
10               0                              Capitol Hill - Downtown Seattle   
                 1                              Capitol Hill - Downtown Seattle   
41               0                  Lake City - Northgate TC - Downtown Seattle   
                 1                  Lake City - Northgate TC - Downtown Seattle   
7                0               Prentice St - Rainier Beach - Downtown Seattle   
                 1               Prentice St - Rainier Beach - Downtown Seattle   
8                0                Seattle Center - Capitol Hill - Rainier Beach   
                 1                Seattle Center - Capitol Hill - Rainier Beach   
C Line           0              Valley St/Fairview Av N - SW Barton St/26 Av SW   
                 1              Valley St/Fairview Av N - SW Barton St/26 Av SW   
E Line           0             Aurora Village Transit Center - Downtown Seattle   
                 1             Aurora Village Transit Center - Downtown Seattle   
Redmond LOOP     0                                                 Redmond Loop   

                                                                   trip_headsign  \
route_short_name direction_id                                                      
10               0                                   Capitol Hill Via 15th Ave E   
                 1               [Downtown Seattle , Atlantic Base Via Broadway]   
41               0             [Lake City Northgate, Northgate Northeast 125t...   
                 1               [Downtown Seattle , Downtown Seattle Northgate]   
7                0             [Rainier Beach Via Rainier Ave S, Prentice St ...   
                 1             [Downtown Seattle Via Rainier Ave S, S Jackson...   
8                0                                               Seattle Center    
                 1                                   Mount Baker Transit Center    
C Line           0             [Westwood Village , West Seattle Alaska Junction]   
                 1                             South Lake Union Downtown Seattle   
E Line           0                                Aurora Village Transit Center    
                 1             [Downtown Seattle , Stadium Station Downtown S...   
Redmond LOOP     0                                                 Redmond Loop    

                               shape_count  trip_count  block_count  \
route_short_name direction_id                                         
10               0                       3         486           36   
                 1                       3         482           36   
41               0                       8         450          110   
                 1                      10         438          102   
7                0                       7         641          137   
                 1                       5         641          137   
8                0                       1         388           67   
                 1                       1         396           67   
C Line           0                       4         579          134   
                 1                       2         581          148   
E Line           0                       1         551          152   
                 1                       2         554          153   
Redmond LOOP     0                       1          18            2   

                              trip_short_name peak_flag                fare_id  
route_short_name direction_id                                                   
10               0                      LOCAL    [0, 1]         [101.0, 110.0]  
                 1                      LOCAL    [0, 1]         [101.0, 110.0]  
41               0                      LOCAL    [0, 1]         [101.0, 110.0]  
                 1                      LOCAL    [0, 1]         [101.0, 110.0]  
7         

In [207]:
gtfs_aug18.trips_by_route_and_direction().shape

(438, 8)

In [217]:
trips_by_direction = gtfs_aug18.trips_by_route_and_direction()
trips_by_direction.loc['101', 'trip_short_name'] == 'LOCAL'
# trips_by_direction[(trips_by_direction.trip_short_name == 'LOCAL')]

direction_id
0    True
1    True
Name: trip_short_name, dtype: bool

In [198]:
gtfs_aug18.trips[gtfs_aug18.trips['fare_id'].isna()].shape

(2027, 10)

In [200]:
gtfs_aug18.trips['fare_id'].unique()

array([101., 110., 100., 501., 305., 900.,  nan, 906., 301., 102., 112.])

In [192]:
len(gtfs_aug18.trips['fare_id'].unique())

11

In [300]:
gtfs_aug18.trips['fare_id'].nunique()

10

In [193]:
gtfs_aug18.trips['fare_id'].unique()[6]

nan

In [194]:
# WTF?! Why is this False??
gtfs_aug18.trips['fare_id'].unique()[6] == np.nan

False

In [203]:
# Oh, you need to use np.isnan
np.isnan(gtfs_aug18.trips['fare_id'].unique()[6])

True

In [299]:
set(gtfs_aug18.trips['fare_id'].head())

{101.0}

In [297]:
gt.list_unique(gtfs_aug18.trips['fare_id'])[:40]

(nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 100.0,
 101.0,
 102.0,
 110.0,
 112.0,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan)

### Double check that our `trip_count` column is counting the right thing

The `count` function will count non-null rows, so our implementation assumes that each row in the trips table represents a unique trip. Let's verify this:

In [130]:
# Count total number of rows in trips table
gtfs_aug18.trips.shape

(59857, 10)

In [125]:
# Count number of unique trip_id's in trips table -- good, it's equal to above
gtfs_aug18.trips['trip_id'].unique().shape

(59857,)

In [123]:
# Make sure our trip counts add up to the right thing
gtfs_aug18.trips_by_route_and_direction()['trip_count'].sum()

59857

## Add another convenience function for the routes table

In [312]:
gtfs_aug18.routes_by_name(7, 120, 'D Line')

route_desc agency_id  \
route_short_name                                                                
120                 Burien TC - Westwood Village - Downtown Seattle       KCM   
7                    Prentice St - Rainier Beach - Downtown Seattle       KCM   
D Line            Blue Ridge/Crown Hill - Ballard - Downtown Sea...       KCM   

                  route_id  
route_short_name            
120                 100019  
7                   100263  
D Line              102581

In [148]:
# If we want a dictionary from route names to route id's. Leave out the route names if you want all routes.
gtfs_aug18.routes_by_name(7, 120, 'D Line')['route_id'].to_dict()

{'120': 100019, '7': 100263, 'D Line': 102581}

In [316]:
gtfs_aug18.fare_attributes

fare_id agency_id  fare_period_id  price        descriptions  \
0       100       KCM               1   0.00    KCM Regular Free   
1       101       KCM               1   2.75  KCM Regular 1 Zone   
2       102       KCM               1   2.75  KCM Regular 2 Zone   
3       110       KCM               2   2.75     KCM Peak 1 Zone   
4       112       KCM               2   2.75     KCM Peak 2 Zone   
5       301       KCM               1   4.25              Boeing   
6       302       KCM               1   3.75              Boeing   
7       305       KCM               1   4.75         Custom High   
8       501        ST               1   2.75   ST Regular 1 Zone   
9       502        ST               1   3.75   ST Regular 2 Zone   
10      700       KCM               1   4.75          Water Taxi   
11      701       KCM               1   5.50   Vashon Water Taxi   
12      900       KCM               1   2.50                DART   
13      902       KCM               1   3.00                LINK   
14      903       KCM               1   2.25                LINK   
15      904       KCM               1   2.50                LINK   
16      905       KCM               1   2.75                LINK   
17      906       KCM               1   3.25                LINK   
18      915       EOS               1   2.25        Stcr Regular   

   currency_type  payment_method  transfers  transfer_duration  
0            USD               0        NaN               5400  
1            USD               0        NaN               5400  
2            USD               0        NaN               5400  
3            USD               0        NaN               5400  
4            USD               0        NaN               5400  
5            USD               0        0.0                  0  
6            USD               0        0.0                  0  
7            USD               0        0.0                  0  
8            USD               0        0.0                  0  
9            USD               0        0.0                  0  
10           USD               0        0.0                  0  
11           USD               0        0.0                  0  
12           USD               0        NaN               5400  
13           USD               0        0.0                  0  
14           USD               0        0.0                  0  
15           USD               0        0.0                  0  
16           USD               1        0.0                  0  
17           USD               1        0.0                  0  
18           USD               1        0.0                  0

## Add function to group by route and shape

Whoa, what's the deal with all the shapes for the 41?

In [209]:
gtfs_aug18.trips_by_route_and_shape(7,8, 41, 'D Line')

direction_id                       trip_headsign  \
route_short_name shape_id                                                     
41               11041008             0                 Lake City Northgate   
                 11041011             0                 Lake City Northgate   
                 11041013             0      Northgate Northeast 125th Only   
                 11041015             0                 Lake City Northgate   
                 11041018             0                 Lake City Northgate   
                 11041020             0                 Lake City Northgate   
                 11041021             0      Northgate Northeast 125th Only   
                 11041023             0                 Lake City Northgate   
                 20041006             1          Downtown Seattle Northgate   
                 20041009             1                   Downtown Seattle    
                 20041010             1                   Downtown Seattle    
                 20041012             1          Downtown Seattle Northgate   
                 20041014             1          Downtown Seattle Northgate   
                 20041016             1          Downtown Seattle Northgate   
                 20041017             1                   Downtown Seattle    
                 20041019             1          Downtown Seattle Northgate   
                 20041022             1          Downtown Seattle Northgate   
                 20041024             1          Downtown Seattle Northgate   
7                10007006             1  Downtown Seattle Via Rainier Ave S   
                 10007007             1  Downtown Seattle Via Rainier Ave S   
                 10007012             1                       S Jackson St    
                 10007014             1  Downtown Seattle Via Rainier Ave S   
                 10007018             1  Downtown Seattle Via Rainier Ave S   
                 21007002             0     Rainier Beach Via Rainier Ave S   
                 21007003             0     Rainier Beach Via Rainier Ave S   
                 21007008             0     Rainier Beach Via Rainier Ave S   
                 21007013             0       Prentice St Via Rainier Ave S   
                 21007015             0       Prentice St Via Rainier Ave S   
                 21007016             0       Prentice St Via Rainier Ave S   
                 21007017             0       Prentice St Via Rainier Ave S   
8                31008004             0                     Seattle Center    
                 40008003             1         Mount Baker Transit Center    
D Line           11674004             0                      Ballard Uptown   
                 20674043             1             Downtown Seattle Uptown   
                 20674045             1    Stadium Station Downtown Seattle   

                           trip_count  block_count trip_short_name peak_flag  \
route_short_name shape_id                                                      
41               11041008          83           37           LOCAL    [0, 1]   
                 11041011         117           37           LOCAL    [0, 1]   
                 11041013          23           19           LOCAL    [0, 1]   
                 11041015           5            5           LOCAL       [0]   
                 11041018          80           32           LOCAL       [0]   
                 11041020         114           35           LOCAL       [0]   
                 11041021          23           19           LOCAL       [0]   
                 11041023           5            5           LOCAL       [0]   
                 20041006         133           38           LOCAL    [0, 1]   
                 20041009          11           11           LOCAL    [0, 1]   
                 20041010           6            6           LOCAL    [0, 1]   
                 20041012          70           36           LO

### Hmm, some of the shape id's changed between July and August

For the 41, there are 18 shapes for both months

In [218]:
gtfs_jul18.trips_by_route_and_shape(7,8, 41, 'D Line')

direction_id                       trip_headsign  \
route_short_name shape_id                                                     
41               11041002             0                 Lake City Northgate   
                 11041008             0                 Lake City Northgate   
                 11041011             0                 Lake City Northgate   
                 11041013             0      Northgate Northeast 125th Only   
                 11041015             0                 Lake City Northgate   
                 11041231             0                 Lake City Northgate   
                 11041232             0                 Lake City Northgate   
                 11041240             0      Northgate Northeast 125th Only   
                 20041003             1                   Downtown Seattle    
                 20041005             1          Downtown Seattle Northgate   
                 20041006             1          Downtown Seattle Northgate   
                 20041009             1                   Downtown Seattle    
                 20041010             1                   Downtown Seattle    
                 20041012             1          Downtown Seattle Northgate   
                 20041014             1          Downtown Seattle Northgate   
                 20041228             1          Downtown Seattle Northgate   
                 20041229             1                   Downtown Seattle    
                 20041250             1          Downtown Seattle Northgate   
7                10007006             1  Downtown Seattle Via Rainier Ave S   
                 10007007             1  Downtown Seattle Via Rainier Ave S   
                 10007012             1                       S Jackson St    
                 10007014             1  Downtown Seattle Via Rainier Ave S   
                 10007018             1  Downtown Seattle Via Rainier Ave S   
                 21007002             0     Rainier Beach Via Rainier Ave S   
                 21007003             0     Rainier Beach Via Rainier Ave S   
                 21007008             0     Rainier Beach Via Rainier Ave S   
                 21007013             0       Prentice St Via Rainier Ave S   
                 21007015             0       Prentice St Via Rainier Ave S   
                 21007016             0       Prentice St Via Rainier Ave S   
                 21007017             0       Prentice St Via Rainier Ave S   
8                31008004             0                     Seattle Center    
                 40008003             1         Mount Baker Transit Center    
D Line           11674004             0                      Ballard Uptown   
                 20674043             1             Downtown Seattle Uptown   
                 20674045             1    Stadium Station Downtown Seattle   

                           trip_count  block_count trip_short_name peak_flag  \
route_short_name shape_id                                                      
41               11041002           5            5           LOCAL       [0]   
                 11041008          83           37           LOCAL    [0, 1]   
                 11041011         117           37           LOCAL    [0, 1]   
                 11041013          23           19           LOCAL    [0, 1]   
                 11041015           5            5           LOCAL       [0]   
                 11041231         114           35           LOCAL       [0]   
                 11041232          80           32           LOCAL       [0]   
                 11041240          23           19           LOCAL       [0]   
                 20041003           6            5           LOCAL       [0]   
                 20041005         128           34           LOCAL       [0]   
                 20041006         133           38           LOCAL    [0, 1]   
                 20041009          11           11           LO

## For some reason the following line originally threw an error

Mon-Tue, March 25-26, 2019

When calling `gtfs_jul18.trips_by_route_and_direction(41)` below, I got the error message "ValueError: Function does not reduce", even though the function seemed to be working fine above. After googling it, the problem [seems to be](https://stackoverflow.com/questions/27439023/pandas-groupby-agg-function-does-not-reduce) that pandas doesn't like it when either (1) an aggregation function can return different types (e.g. int or list), or (2) the returned types are not hashable (e.g. a list vs. a tuple).

The recommended solution was to always return a tuple, which is immutable and hence hashable, if you want to aggregate into a list. (Strangely, calling `.nunique()` sometimes gave the same "Function does not reduce" error as `.unique()`, even though it should always just return a number, not a numpy array. I think the problem has to do with the `nan`'s.)

I wrote a collection of simple aggregation functions such as `def list_unique(hashable_objects): return tuple(set(hashable_objects))` and used these instead of calling `.unique()` or `.nunique()`. So far, this seems to work...

In [314]:
gtfs_jul18.trips_by_route_and_direction(41)

route_desc  \
route_short_name direction_id                                                
41               0             Lake City - Northgate TC - Downtown Seattle   
                 1             Lake City - Northgate TC - Downtown Seattle   

                                                                   trip_headsign  \
route_short_name direction_id                                                      
41               0             (Lake City Northgate, Northgate Northeast 125t...   
                 1               (Downtown Seattle , Downtown Seattle Northgate)   

                               shape_count  trip_count  block_count  \
route_short_name direction_id                                         
41               0                       8         450          110   
                 1                      10         438          102   

                              trip_short_name peak_flag  \
route_short_name direction_id                             
41               0                   (LOCAL,)    (0, 1)   
                 1                   (LOCAL,)    (0, 1)   

                                                   fare_descriptions  
route_short_name direction_id                                         
41               0             (KCM Peak 1 Zone, KCM Regular 1 Zone)  
                 1             (KCM Peak 1 Zone, KCM Regular 1 Zone)

In [315]:
# WTF?! Why does this work now even though I haven't updated the aggregation functions?
gtfs_jul18.trips_by_route_and_shape(41)

direction_id                      trip_headsign  \
route_short_name shape_id                                                    
41               11041002             0             (Lake City Northgate,)   
                 11041008             0             (Lake City Northgate,)   
                 11041011             0             (Lake City Northgate,)   
                 11041013             0  (Northgate Northeast 125th Only,)   
                 11041015             0             (Lake City Northgate,)   
                 11041231             0             (Lake City Northgate,)   
                 11041232             0             (Lake City Northgate,)   
                 11041240             0  (Northgate Northeast 125th Only,)   
                 20041003             1               (Downtown Seattle ,)   
                 20041005             1      (Downtown Seattle Northgate,)   
                 20041006             1      (Downtown Seattle Northgate,)   
                 20041009             1               (Downtown Seattle ,)   
                 20041010             1               (Downtown Seattle ,)   
                 20041012             1      (Downtown Seattle Northgate,)   
                 20041014             1      (Downtown Seattle Northgate,)   
                 20041228             1      (Downtown Seattle Northgate,)   
                 20041229             1               (Downtown Seattle ,)   
                 20041250             1      (Downtown Seattle Northgate,)   

                           trip_count  block_count trip_short_name peak_flag  \
route_short_name shape_id                                                      
41               11041002           5            5        (LOCAL,)      (0,)   
                 11041008          83           37        (LOCAL,)    (0, 1)   
                 11041011         117           37        (LOCAL,)    (0, 1)   
                 11041013          23           19        (LOCAL,)    (0, 1)   
                 11041015           5            5        (LOCAL,)      (0,)   
                 11041231         114           35        (LOCAL,)      (0,)   
                 11041232          80           32        (LOCAL,)      (0,)   
                 11041240          23           19        (LOCAL,)      (0,)   
                 20041003           6            5        (LOCAL,)      (0,)   
                 20041005         128           34        (LOCAL,)      (0,)   
                 20041006         133           38        (LOCAL,)    (0, 1)   
                 20041009          11           11        (LOCAL,)    (0, 1)   
                 20041010           6            6        (LOCAL,)    (0, 1)   
                 20041012          70           36        (LOCAL,)    (0, 1)   
                 20041014           2            2        (LOCAL,)      (0,)   
                 20041228          69           32        (LOCAL,)      (0,)   
                 20041229          11           11        (LOCAL,)      (0,)   
                 20041250           2            2        (LOCAL,)      (0,)   

                                               fare_descriptions  
route_short_name shape_id                                         
41               11041002                  (KCM Regular 1 Zone,)  
                 11041008  (KCM Peak 1 Zone, KCM Regular 1 Zone)  
                 11041011  (KCM Peak 1 Zone, KCM Regular 1 Zone)  
                 11041013  (KCM Peak 1 Zone, KCM Regular 1 Zone)  
                 11041015                  (KCM Regular 1 Zone,)  
                 11041231                  (KCM Regular 1 Zone,)  
                 11041232                  (KCM Regular 1 Zone,)  
                 11041240                  (KCM Regular 1 Zone,)  
                 20041003                  (KCM Regular 1 Zone,)  
                 20041005                  (KCM Regular 1 Zone,)  
                 20041006  (KCM Peak 1 Zone, KCM Regular 1 Zo

In [277]:
gtfs_jul18.route_ids()['41']
# gtfs_jul18.trips.set_index('route_id').loc[gtfs_jul18.route_ids(41, 'C Line')]

100221

In [273]:
# gtfs_jul18.routes_by_name(41)['route_id'].to_dict()['41']
# gtfs_jul18.routes_by_name().loc['41', 'route_id']
# jul_trips_41 = gtfs_jul18.trips.set_index('route_id').loc[gtfs_jul18.routes_by_name().loc['41', 'route_id']]
jul_trips_41 = gtfs_jul18.trips.set_index('route_id').loc[gtfs_jul18.route_ids(41)]
jul_trips_41.head(10)

service_id   trip_id               trip_headsign trip_short_name  \
route_id                                                                     
100221         44711  39485224           Downtown Seattle            LOCAL   
100221         44711  39485225         Lake City Northgate           LOCAL   
100221         44711  39485226  Downtown Seattle Northgate           LOCAL   
100221         44711  39485777         Lake City Northgate           LOCAL   
100221         44711  39485778  Downtown Seattle Northgate           LOCAL   
100221         44711  39485779  Downtown Seattle Northgate           LOCAL   
100221         44711  39485780  Downtown Seattle Northgate           LOCAL   
100221         44711  39485781         Lake City Northgate           LOCAL   
100221         44711  39485782  Downtown Seattle Northgate           LOCAL   
100221         44711  39485783  Downtown Seattle Northgate           LOCAL   

          direction_id  block_id  shape_id  peak_flag  fare_id  
route_id                                                        
100221               1   5062060  20041003          0    101.0  
100221               0   5061877  11041232          0    101.0  
100221               1   5061877  20041228          0    101.0  
100221               0   5061877  11041232          0    101.0  
100221               1   5061873  20041005          0    101.0  
100221               1   5061872  20041005          0    101.0  
100221               1   5061871  20041005          0    101.0  
100221               0   5061871  11041231          0    101.0  
100221               1   5061880  20041228          0    101.0  
100221               1   5061862  20041005          0    101.0

In [274]:
jul_trips_41.shape

(888, 9)

In [242]:
jul_trips_41.fare_id.unique()

array([101., 110.])

In [243]:
jul_trips_41.fare_id.nunique()

2

In [249]:
jul_trips_41.groupby('direction_id').agg({'fare_id': lambda x: x.nunique()})

fare_id
direction_id         
0                 2.0
1                 2.0

In [287]:
sorted([gtfs_aug18, gtfs_jul18, gtfs_aug18])

TypeError: '<' not supported between instances of 'StaticGTFS' and 'StaticGTFS'